File: Sliding Window/3. Longest Substring Without Repeating Characters/Solution.java
import java.lang.Math;
import java.util.*;

class Solution {
    public int lengthOfLongestSubstring(String s) {
        // N : len(s), M : Longest unique SubString
        // Time Complexity: O(N^2)
        // Space Complexity: O(M)
        // int n = s.length();
        // HashSet<Character> charSet = new HashSet<>();
        // int longestString = 0;
        // for (int i=0; i<n; i++){
        //     charSet = new HashSet<>();
        //     for (int j=i; j<n; j++){
        //         if (charSet.contains(s.charAt(j))){
        //             longestString = Math.max(longestString, j - i);
        //             break;
        //         }
        //         charSet.add(s.charAt(j));
        //     }
        //     longestString = Math.max(longestString, charSet.size());
        // }
        // return longestString;


        // N : len(s), M : Longest unique SubString
        // Time Complexity: O(N)
        // Space Complexity: O(M)
        int n = s.length();
        HashSet<Character> charSet = new HashSet<>();
        int longestString = 0;
        int l = 0;
        int r = 0;
        while (r < n){
            if (charSet.contains(s.charAt(r))){
                longestString = Math.max(longestString, r - l);
                charSet.remove(s.charAt(l));
                l++;
                continue;
            }
            charSet.add(s.charAt(r));
            r++;
        }
        return longestString = Math.max(longestString, r - l);
    }
}

File: Sliding Window/3. Longest Substring Without Repeating Characters/Solution.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        # N : len(s), M : Longest unique SubString
        # Time Complexity: O(N^2)
        # Space Complexity: O(M)
        # n = len(s)
        # longestString = 0
        # for i in range(n):
        #     charSet = set()
        #     for j in range(i, n):
        #         if s[j] in charSet:
        #             longestString = max(longestString, j - i)
        #             break
        #         charSet.add(s[j])
        #     longestString = max(longestString, len(charSet))
        # return longestString


        # N : len(s), M : Longest unique SubString
        # Time Complexity: O(N)
        # Space Complexity: O(M)
        n = len(s)
        charSet = set()
        longestString = 0
        l, r = 0, 0
        while r < n:
            if s[r] in charSet:
                longestString = max(longestString, r - l)
                charSet.remove(s[l])
                l += 1
                continue
            charSet.add(s[r])
            r += 1
        longestString = max(longestString, r - l)
        return longestString